### PR TITLE
use string buffer to store generated code

### PR DESCRIPTION
**Pull request by StrikerRUS**
While changes in #207 are arguably giving significant improvement, the speedup achieved by these updates can be seen on transpiling medium-size models without any profiling tools.

During skimming the report from https://github.com/BayesWitnesses/m2cgen/pull/207#issuecomment-625898131, I noticed that `add_code_line()` takes a huge amount of time (also refer to #209 for another strings-related improvement).

Here is a simple bench.

```
import sys

from sklearn.datasets import load_boston

import lightgbm as lgb
import m2cgen as m2c

X, y = load_boston(True)
est = lgb.LGBMRegressor(n_estimators=1000, random_state=42).fit(X, y)

sys.setrecursionlimit(int(1e5))

%%timeit -n7 -r5
_ = m2c.export_to_python(est)
```

And the results are the following:
```
30.6 s ± 16.9 ms per loop (mean ± std. dev. of 5 runs, 7 loops each)
```
```
2.41 s ± 4.84 ms per loop (mean ± std. dev. of 5 runs, 7 loops each)
```

### CODE REVIEW

File: m2cgen/interpreters/code_generator.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
+from io import StringIO
 from string import Template
+from weakref import finalize
```
Comment (StrikerRUS) [inline m2cgen/interpreters/code_generator.py:3]: https://docs.python.org/3/library/weakref.html#comparing-finalizers-with-del-methods

File: m2cgen/interpreters/code_generator.py [hunk 2]
```diff
@@ -29,11 +31,32 @@ class BaseCodeGenerator:
...
+        self._code_buf = None
         self.reset_state()
 
     def reset_state(self):
         self._current_indent = 0
-        self.code = ""
+        self._finalize_buffer()
+        self._code_buf = StringIO()
+        self._code = None
+        self._finalizer = finalize(self, self._finalize_buffer)
```
Comment (StrikerRUS) [inline m2cgen/interpreters/code_generator.py:42]: Great post! https://forum.omz-software.com/topic/4473/python-question-class-cleanup-file-handles-etc/7

**Comment by izeigerman** [comment]
Oh, wow. This is a dramatic improvement indeed! Fantastic job 👍

File: m2cgen/interpreters/code_generator.py (hunk 2 above)
Comment (izeigerman) [reply m2cgen/interpreters/code_generator.py:42]: Nice, yet another Python TIL for me!

File: m2cgen/interpreters/code_generator.py [hunk 3]
```diff
@@ -29,11 +31,32 @@ class BaseCodeGenerator:
...
+            self._code_buf.close()
+
+    def _check_buf_closed(self):
+        if self._code_buf.closed:
+            raise BufferError(
+                "Cannot modify code after getting generated code and "
+                "closing the underlying buffer!\n"
+                "Call reset_state() to allocate new buffer.")
+
+    def get_generated_code(self):
```
[the remark this answers is not in the document]
Comment (izeigerman) [reply m2cgen/interpreters/code_generator.py:55]: That being said we don't necessarily have to support repeated invocation for this method and throw an exception on subsequent method calls. Please correct me if I'm wrong here.

File: m2cgen/interpreters/code_generator.py [hunk 4]
```diff
@@ -29,11 +31,32 @@ class BaseCodeGenerator:
...
+        if not self._code_buf.closed:
+            self._code = self._code_buf.getvalue()
+            self._finalize_buffer()
+        return self._code if self._code is not None else ""
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply m2cgen/interpreters/code_generator.py:59]: I always prefer to check for `None` explicitly according to the recommendation from PEP.
https://github.com/BayesWitnesses/m2cgen/blob/2c2b17453eba1ad36c909c1b97fd1a8ca6570f97/m2cgen/interpreters/code_generator.py#L41

> Also, beware of writing if x when you really mean if x is not None -- e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!
  https://www.python.org/dev/peps/pep-0008/#programming-recommendations

> If you only used if key: here, then an argument which evaluated to false would not be considered. Explicitly comparing with is None is the correct idiom to make this check. See Truth Value Testing.
   https://stackoverflow.com/a/17117361

https://stackoverflow.com/a/7816439

File: m2cgen/interpreters/code_generator.py [hunk 5]
```diff
@@ -29,11 +31,32 @@ class BaseCodeGenerator:
...
+
+    def _check_buf_closed(self):
+        if self._code_buf.closed:
+            raise BufferError(
+                "Cannot modify code after getting generated code and "
+                "closing the underlying buffer!\n"
+                "Call reset_state() to allocate new buffer.")
+
+    def get_generated_code(self):
+        if not self._code_buf.closed:
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply m2cgen/interpreters/code_generator.py:56]: Nope. We call `reset_state()` in `__init__()` where `_code_fuf` is set to `StringIO()`. So we have a guarantee here that `_code_buf` is already was created (not `None` from `__init__()`).

File: m2cgen/interpreters/code_generator.py [hunk 6]
```diff
@@ -46,24 +69,34 @@ def decrease_indent(self):
     # All code modifications should be implemented via following methods.
 
     def add_code_line(self, line):
+        self._check_buf_closed()
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply m2cgen/interpreters/code_generator.py:72]: Sound good! I'll do it right now.

File: m2cgen/interpreters/code_generator.py [hunk 7]
```diff
@@ -46,24 +69,34 @@ def decrease_indent(self):
...
         if isinstance(lines, str):
             lines = lines.strip().split("\n")
         indent = " " * self._current_indent
-        self.code += indent + "\n{}".format(indent).join(lines) + "\n"
+        self._code_buf.write(indent + "\n{}".format(indent).join(lines) + "\n")
 
     def prepend_code_line(self, line):
-        self.code = line + "\n" + self.code
+        self._check_buf_closed()
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply m2cgen/interpreters/code_generator.py:86]: Nice catch!

File: m2cgen/interpreters/code_generator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply m2cgen/interpreters/code_generator.py:55]: Absolutely agree for changing the name!

> That being said we don't necessarily have to support repeated invocations for this method and throw an exception on subsequent method calls. Please correct me if I'm wrong here.

That's true but only for now. Maybe in the future we will need something like 

```
if cg.get_generated_code() != "":
    return cg.get_generated_code()
else:
    ...
```

Of course, we can save the result from the first call into a variable, but I don't see any reasons to eliminate the opportunity to return the same value from this method in its subsequent calls. I believe it is in the consistency of functional programming paradigm 🙂 .

**Comment by StrikerRUS** [comment]
Thanks a lot for the thoughtful review! I addressed review comments in the two latest commits and left some replies. Please give it another round of review.

**Comment by izeigerman** [comment]
@StrikerRUS does this PR require any additional updates considering that the Ruby PR has been merged?

**Comment by StrikerRUS** [comment]
> does this PR require any additional updates considering that the Ruby PR has been merged?

Yeah, definitely! Just pushed the needed changes.

**Pull request merged**